{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f68fd5ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "import geopandas as gpd, pandas as pd, matplotlib.pyplot as plt, plotly, numpy as np\n",
    "\n",
    "import plotly.express as px \n",
    "import plotly.graph_objects as go\n",
    "import pandas_datareader.data as web\n",
    "import datetime\n",
    "import dash\n",
    "import dash_html_components as html\n",
    "import dash_core_components as dcc\n",
    "import plotly.graph_objects as go\n",
    "import dash_bootstrap_components as dbc\n",
    "from dash.dependencies import Input, Output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ac80bf7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "points_within = gpd.read_file('/Users/magio94/Dropbox/Bottom-up/_Tests_area/Tests_ZukunftsmesseWL/Shapefiles/points_within.shp')\n",
    "municipalities_westfalen_lippe = gpd.read_file('/Users/magio94/Dropbox/Bottom-up/_Tests_area/Tests_ZukunftsmesseWL/Shapefiles/municipalities_westfalen_lippe.shp')\n",
    "#activity_count = gpd.read_file('/Users/magio94/Dropbox/Bottom-up/_Tests_area/Tests_ZukunftsmesseWL/Shapefiles/activity_count.shp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "97950be3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash is running on http://127.0.0.1:8000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:8000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:8000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:8000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:8000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:8000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:8000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:8000/\n",
      "\n",
      " * Serving Flask app '__main__' (lazy loading)\n",
      " * Environment: production\n",
      "\u001b[31m   WARNING: This is a development server. Do not use it in a production deployment.\u001b[0m\n",
      "\u001b[2m   Use a production WSGI server instead.\u001b[0m\n",
      " * Debug mode: off\n"
     ]
    }
   ],
   "source": [
    "dff = points_within.copy()\n",
    "dff2 = municipalities_westfalen_lippe.copy()\n",
    "\n",
    "# https://www.bootstrapcdn.com/bootswatch/\n",
    "app = dash.Dash(__name__, external_stylesheets=[dbc.themes.CYBORG],\n",
    "                meta_tags=[{'name': 'viewport',\n",
    "                            'content': 'width=device-width, initial-scale=1.0'}]\n",
    "                )\n",
    "\n",
    "app.layout = dbc.Container([\n",
    "\n",
    "    dbc.Row(\n",
    "        dbc.Col(html.H1(\"Different activities by category\",\n",
    "                        className='text-center bg-gradient-primary text-white mb-4'),\n",
    "                width=12)\n",
    "    ),\n",
    "    \n",
    "    dbc.Row([\n",
    "        dbc.Col([\n",
    "            html.H4(\"Map divided by categorical activities\",\n",
    "                    className='text-center bg-gradient-primary text-white mb-4'),\n",
    "            \n",
    "            dcc.Dropdown(id=\"category_selector\",\n",
    "                         options=[\n",
    "                             {'label': i, 'value': i} for i in dff['category'].unique()],\n",
    "                         value=['Association', 'NPO'],\n",
    "                         multi=True,\n",
    "                             ),\n",
    "            dcc.Graph(id='map')],\n",
    "            width={'size':5, \n",
    "                   'offset':1\n",
    "                  }),       \n",
    "    \n",
    "        dbc.Col([dcc.Dropdown(id=\"dropdown\",\n",
    "                              options=[\n",
    "                                  {'label': n, 'value': n} for n in dff['NAME_2'].unique()],\n",
    "                              value = 'Arnsberg',\n",
    "                              multi=False,\n",
    "                             ),\n",
    "                \n",
    "                dcc.Graph(id='bar-chart')],\n",
    "                className= 'text-center bg-secondary text-white mb-4',\n",
    "                width={'size':5, \n",
    "                       #'offset':1\n",
    "                      }),      \n",
    "    ]),\n",
    "    \n",
    "], \n",
    "\n",
    "    fluid=True)\n",
    "\n",
    "@app.callback(Output('map', 'figure'),\n",
    "              [Input('category_selector', 'value')])\n",
    "\n",
    "def update_graph(option_slctd):\n",
    "    dff = points_within.copy()\n",
    "    \n",
    "    dff = dff[dff['category'].isin(option_slctd)]\n",
    "    dff2 = municipalities_westfalen_lippe.copy()\n",
    "\n",
    "    fig = px.choropleth_mapbox(dff2,\n",
    "                               geojson=dff2.geometry,\n",
    "                               locations=dff2.index,\n",
    "                               color=\"WL_provinc\",\n",
    "                               mapbox_style=\"open-street-map\",\n",
    "                               opacity=0.5,\n",
    "                               zoom=6,\n",
    "                               center = {\"lat\": 51.71027, \"lon\": 8.059108}\n",
    "                              )\n",
    "    fig.update(layout_showlegend=False)\n",
    "\n",
    "    fig2 = px.scatter_mapbox(dff,\n",
    "                             lat=\"Lat\",\n",
    "                             lon=\"Lon\",\n",
    "                             # hover_name= \"name\",\n",
    "                             # color_discrete_sequence=[\"fuchsia\"],\n",
    "                             color=dff.category,\n",
    "                             # hover_data = [\"name\"]\n",
    "                             )\n",
    "\n",
    "    fig2.update_layout(mapbox_style=\"white-bg\")\n",
    "    fig.update(layout_showlegend=True)\n",
    "\n",
    "    for x in range(0, len(option_slctd)):\n",
    "        fig.add_trace(fig2.data[x])\n",
    "\n",
    "\n",
    "    return fig\n",
    "\n",
    "@app.callback(\n",
    "    Output(\"bar-chart\", \"figure\"), \n",
    "    [Input(\"dropdown\", \"value\")])\n",
    "\n",
    "def update_bar_chart(NAME_2):\n",
    "    \n",
    "    dff2 = dff[dff['NAME_2'] == NAME_2]\n",
    "    \n",
    "    fig = px.bar(dff2.category, y= dff2.NAME_3)\n",
    "    fig.update_layout(yaxis={'categoryorder':'total descending'})\n",
    "    \n",
    "    return fig\n",
    "\n",
    "if __name__=='__main__':\n",
    "    app.run_server(debug=False, port=8000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "6bdbff07",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:3000/\n",
      "\n",
      " * Serving Flask app '__main__' (lazy loading)\n",
      " * Environment: production\n",
      "\u001b[31m   WARNING: This is a development server. Do not use it in a production deployment.\u001b[0m\n",
      "\u001b[2m   Use a production WSGI server instead.\u001b[0m\n",
      " * Debug mode: off\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n",
      "Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 2051, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1501, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1499, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\", line 1485, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1096, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "  File \"/Users/magio94/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\", line 1017, in add_context\n",
      "    output_value = func(*args, **kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-126-51aeb8b0ae8c>\", line 44, in update_graph\n",
      "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
      "AttributeError: 'list' object has no attribute 'split'\n"
     ]
    }
   ],
   "source": [
    "dff = points_within.copy()\n",
    "\n",
    "# https://www.bootstrapcdn.com/bootswatch/\n",
    "app = dash.Dash(__name__, external_stylesheets=[dbc.themes.CYBORG],\n",
    "                meta_tags=[{'name': 'viewport',\n",
    "                            'content': 'width=device-width, initial-scale=1.0'}]\n",
    "                )\n",
    "\n",
    "app.layout = dbc.Container([\n",
    "\n",
    "    dbc.Row(\n",
    "        dbc.Col(html.H1(\"Different activities by category\",\n",
    "                        className='text-center bg-gradient-primary text-white mb-4'),\n",
    "                width=12)\n",
    "    ),\n",
    "    \n",
    "    dbc.Row([\n",
    "        dbc.Col([\n",
    "            html.H4(\"Map divided by categorical activities\",\n",
    "                    className='text-center bg-gradient-primary text-white mb-4'),\n",
    "            \n",
    "            dcc.Dropdown(id=\"category_selector\",\n",
    "                         options=[\n",
    "                             {'label': i, 'value': i} for i in dff['category'].unique()],\n",
    "                         value = 'Association',\n",
    "                         multi=True,\n",
    "                             ),\n",
    "            dcc.Graph(id='map')],\n",
    "            width={'size':5, \n",
    "                   'offset':1\n",
    "                  }),       \n",
    "    ]),\n",
    "    \n",
    "], \n",
    "\n",
    "    fluid=True)\n",
    "\n",
    "@app.callback(Output('map', 'figure'),\n",
    "              [Input('category_selector', 'value')])\n",
    "\n",
    "def update_graph(option_slctd):\n",
    "    dff = points_within.copy()\n",
    "    dff = dff[dff['category'].isin(list(option_slctd.split()))]\n",
    "\n",
    "    fig2 = px.scatter_mapbox(dff,\n",
    "                             lat=\"Lat\", \n",
    "                             lon=\"Lon\", \n",
    "                             #hover_name= \"name\",\n",
    "                             #color_discrete_sequence=[\"fuchsia\"],\n",
    "                             color = dff.category,\n",
    "                             #hover_data = [\"name\"]\n",
    "                            )\n",
    "    \n",
    "    fig2.update_layout(mapbox_style=\"white-bg\")\n",
    "     \n",
    "    return fig2\n",
    "\n",
    "\n",
    "if __name__=='__main__':\n",
    "    app.run_server(debug=False, port=3000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6991e6ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash is running on http://127.0.0.1:4000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:4000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:4000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:4000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:4000/\n",
      "\n",
      "Dash is running on http://127.0.0.1:4000/\n",
      "\n",
      " * Serving Flask app '__main__' (lazy loading)\n",
      " * Environment: production\n",
      "\u001b[31m   WARNING: This is a development server. Do not use it in a production deployment.\u001b[0m\n",
      "\u001b[2m   Use a production WSGI server instead.\u001b[0m\n",
      " * Debug mode: off\n"
     ]
    },
    {
     "ename": "OSError",
     "evalue": "[Errno 48] Address already in use",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-f40d0d43f587>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     78\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     79\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0m__name__\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'__main__'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 80\u001b[0;31m     \u001b[0mapp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_server\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdebug\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mport\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m4000\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/dash/dash.py\u001b[0m in \u001b[0;36mrun_server\u001b[0;34m(self, host, port, proxy, debug, dev_tools_ui, dev_tools_props_check, dev_tools_serve_dev_bundles, dev_tools_hot_reload, dev_tools_hot_reload_interval, dev_tools_hot_reload_watch_interval, dev_tools_hot_reload_max_retry, dev_tools_silence_routes_logging, dev_tools_prune_errors, **flask_run_options)\u001b[0m\n\u001b[1;32m   1748\u001b[0m                     \u001b[0mextra_files\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1749\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1750\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mserver\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhost\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mhost\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mport\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mport\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdebug\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdebug\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mflask_run_options\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/flask/app.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(self, host, port, debug, load_dotenv, **options)\u001b[0m\n\u001b[1;32m    918\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    919\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 920\u001b[0;31m             \u001b[0mrun_simple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcast\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhost\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mport\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    921\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    922\u001b[0m             \u001b[0;31m# reset the first request information if the development server\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/werkzeug/serving.py\u001b[0m in \u001b[0;36mrun_simple\u001b[0;34m(hostname, port, application, use_reloader, use_debugger, use_evalex, extra_files, exclude_patterns, reloader_interval, reloader_type, threaded, processes, request_handler, static_files, passthrough_errors, ssl_context)\u001b[0m\n\u001b[1;32m   1006\u001b[0m         )\n\u001b[1;32m   1007\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1008\u001b[0;31m         \u001b[0minner\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1009\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1010\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/werkzeug/serving.py\u001b[0m in \u001b[0;36minner\u001b[0;34m()\u001b[0m\n\u001b[1;32m    946\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mLookupError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    947\u001b[0m             \u001b[0mfd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 948\u001b[0;31m         srv = make_server(\n\u001b[0m\u001b[1;32m    949\u001b[0m             \u001b[0mhostname\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    950\u001b[0m             \u001b[0mport\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/werkzeug/serving.py\u001b[0m in \u001b[0;36mmake_server\u001b[0;34m(host, port, app, threaded, processes, request_handler, passthrough_errors, ssl_context, fd)\u001b[0m\n\u001b[1;32m    778\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"cannot have a multithreaded and multi process server.\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    779\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mthreaded\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 780\u001b[0;31m         return ThreadedWSGIServer(\n\u001b[0m\u001b[1;32m    781\u001b[0m             \u001b[0mhost\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mport\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mapp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrequest_handler\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpassthrough_errors\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mssl_context\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfd\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfd\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    782\u001b[0m         )\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/werkzeug/serving.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, host, port, app, handler, passthrough_errors, ssl_context, fd)\u001b[0m\n\u001b[1;32m    684\u001b[0m                 \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munlink\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mserver_address\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    685\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 686\u001b[0;31m         \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mserver_address\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhandler\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# type: ignore\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    687\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    688\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mapp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/socketserver.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, server_address, RequestHandlerClass, bind_and_activate)\u001b[0m\n\u001b[1;32m    450\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mbind_and_activate\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    451\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 452\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mserver_bind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    453\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mserver_activate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    454\u001b[0m             \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/http/server.py\u001b[0m in \u001b[0;36mserver_bind\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    136\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mserver_bind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    137\u001b[0m         \u001b[0;34m\"\"\"Override server_bind to store the server name.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 138\u001b[0;31m         \u001b[0msocketserver\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTCPServer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mserver_bind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    139\u001b[0m         \u001b[0mhost\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mport\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mserver_address\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    140\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mserver_name\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msocket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetfqdn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhost\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/socketserver.py\u001b[0m in \u001b[0;36mserver_bind\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    464\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mallow_reuse_address\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    465\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msocket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msetsockopt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msocket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSOL_SOCKET\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msocket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSO_REUSEADDR\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 466\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msocket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mserver_address\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    467\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mserver_address\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msocket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetsockname\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    468\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mOSError\u001b[0m: [Errno 48] Address already in use"
     ]
    }
   ],
   "source": [
    "points_within = gpd.read_file('/Users/magio94/Dropbox/Bottom-up/_Tests_area/Tests_ZukunftsmesseWL/Shapefiles/points_within.shp')\n",
    "municipalities_westfalen_lippe = gpd.read_file('/Users/magio94/Dropbox/Bottom-up/_Tests_area/Tests_ZukunftsmesseWL/Shapefiles/municipalities_westfalen_lippe.shp')\n",
    "\n",
    "dff = points_within.copy()\n",
    "\n",
    "# https://www.bootstrapcdn.com/bootswatch/\n",
    "app = dash.Dash(__name__, external_stylesheets=[dbc.themes.CYBORG],\n",
    "                meta_tags=[{'name': 'viewport',\n",
    "                            'content': 'width=device-width, initial-scale=1.0'}]\n",
    "                )\n",
    "\n",
    "app.layout = dbc.Container([\n",
    "\n",
    "    dbc.Row(\n",
    "        dbc.Col(html.H1(\"Different activities by category\",\n",
    "                        className='text-center bg-gradient-primary text-white mb-4'),\n",
    "                width=12)\n",
    "    ),\n",
    "\n",
    "    dbc.Row([\n",
    "        dbc.Col([\n",
    "            html.H4(\"Map divided by categorical activities\",\n",
    "                    className='text-center bg-gradient-primary text-white mb-4'),\n",
    "\n",
    "            dcc.Dropdown(id=\"category_selector\",\n",
    "                         options=[\n",
    "                             {'label': i, 'value': i} for i in dff['category'].unique()],\n",
    "                         value=['Association', 'NPO'],\n",
    "                         multi=True,\n",
    "                         ),\n",
    "            dcc.Graph(id='map')],\n",
    "            width={'size': 5,\n",
    "                   'offset': 1\n",
    "                   }),\n",
    "    ]),\n",
    "\n",
    "],\n",
    "\n",
    "    fluid=True)\n",
    "\n",
    "\n",
    "@app.callback(Output('map', 'figure'),\n",
    "              [Input('category_selector', 'value')])\n",
    "def update_graph(option_slctd):\n",
    "    dff = points_within.copy()\n",
    "    dff = dff[dff['category'].isin(option_slctd)]\n",
    "    dff2 = municipalities_westfalen_lippe.copy()\n",
    "\n",
    "    fig = px.choropleth_mapbox(dff2,\n",
    "                               geojson=dff2.geometry,\n",
    "                               locations=dff2.index,\n",
    "                               color=\"WL_provinc\",\n",
    "                               mapbox_style=\"open-street-map\",\n",
    "                               opacity=0.5,\n",
    "                               zoom=6,\n",
    "                               center = {\"lat\": 51.71027, \"lon\": 8.059108}\n",
    "                              )\n",
    "    fig.update(layout_showlegend=False)\n",
    "\n",
    "    fig2 = px.scatter_mapbox(dff,\n",
    "                             lat=\"Lat\",\n",
    "                             lon=\"Lon\",\n",
    "                             # hover_name= \"name\",\n",
    "                             # color_discrete_sequence=[\"fuchsia\"],\n",
    "                             color=dff.category,\n",
    "                             # hover_data = [\"name\"]\n",
    "                             )\n",
    "\n",
    "    fig2.update_layout(mapbox_style=\"white-bg\")\n",
    "    fig.update(layout_showlegend=True)\n",
    "\n",
    "    for x in range(0, len(option_slctd)):\n",
    "        fig.add_trace(fig2.data[x])\n",
    "\n",
    "\n",
    "    return fig\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    app.run_server(debug=False, port=4000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e917d84e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
